### DISCUSSION TITLE
FastAPI requests to internal endpoint

### DISCUSSION THREAD

**Discussion by jayycarter** [fastapi/fastapi#6946] [Questions] [answered]
I was working on a project recently and was experiencing problems with FastAPI interacting with the requests library. The goal I’m trying to achieve is to have a health_status endpoint function hit the health version endpoint as well. I have recreated the problem with 2 smaller projects, a simple fastapi and a simple flask. Flask is able to achieve the goal, but Fastapi freezes up when trying to do it.

Here is the code using flask. When hitting the “/health/status” endpoin it should return {“message”: “Hello World”} however it freezes up the entire api. It doesn’t even allow me to hit the root endpoint afterwards unless I rerun the program. However, I am able to hit the root endpoint if I hit it before calling “/health/status."

### Code

```
import uvicorn
from fastapi import FastAPI, Request
import requests

app = FastAPI(debug=True)


@app.get("/")
async def root():
    return {"message": "Hello World"}


@app.get("/health/status")
async def health():
    resp = requests.get(url="http://127.0.0.1:8000/")
    return resp.text


if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8000)

```

Similarly, the code using flask is listed below. This works just fine, where if I hit “/health/status” then it will return “Hello World”

```
from flask import Flask
import requests

app = Flask(__name__)


@app.route('/')
def hello_world():
    return 'Hello World'


@app.route('/health/status')
def health_status():
    resp = requests.get(url="http://127.0.0.1:5000/")
    return resp.text


if __name__ == '__main__':
    app.run()

```

### Description

As mentioned above, instead of returning "Hello World" like it should, it just ties up the whole api when using FastAPI.

### Environment

* OS: [e.g. Linux / Windows / macOS]: Windows
* FastAPI Version [e.g. 0.3.0]: "==0.63.0"
* Python version: "==3.6.8"

**Comment by ycd** [comment]
> When hitting the “/health/status” endpoin it should return {“message”: “Hello World”} however it freezes up the entire api.

You are executing a synchronous function inside a coroutine, of course, it is going to block. 

But it shouldn't take longer than >100ms.

**Comment by ArcLightSlavik** [comment]
Use `httpx` instead of `requests`, would make your call asynchronous and stop the blocking.

```
async with httpx.AsyncClient() as client:
    resp = await client.get(url="http://127.0.0.1:8000/")
```

**Comment by Dustyposa** [comment]
You can use normal function to solve now.
```python
@app.get("/health/status")
def health():
    resp = requests.get(url="http://127.0.0.1:8000/")
    return resp.text
```

**Comment by Kludex** [comment, marked as answer]
The thing is that you're calling a blocking synchronous function on a `coroutine` and it will block the event loop. Possible solutions are:   
1. Make `health()` synchronous i.e. remove the `async`. This way it will run in another thread.
2. Use the @ArcLightSlavik approach, meaning that you await for the response. In other words, you give time on the event loop to other functions (`root()` in this case). 

Approach 2 performs better.

**Comment by jetsbee** [comment]
@Kludex Can I get to know more details the reason? why the server gets freezing infinitely as requesting "/health/status" in original code from @jayycarter.

I'm wondering why "requests.get()" blocks "async def health()" INFINITELY, not just block some time.

**Comment by jetsbee** [comment]
I think the reason infinite freezing is because "async def health()" is coroutine.
When "requests.get(url="http://127.0.0.1:8000/")" executed, "async def health()" is hanging on for finishing "request.get()".
But it's going to be never finished. "async def health()" is coroutine and it's hanging, so fastapi can't get request for "/" and never invoke "root()".
